ClinGen/clincoded
# Connects to #290, Connects to #507 unminified js with maps plus bundle and css rename

Addresses: 
- #290: Write js.map files even if JS is not minified 
- #507: force bundled JS and compiled css file refresh on new releases

Before testing:
`rm -rf node_modules`


**To test #290:**
Dev (un-minified JS bundle but still has map files):
- Remove bundled JS and associated files:  `rm src/clincoded/static/build/*js*`
- Install npm packages re-run grunt: `bin/buildout -c buildout-dev.cfg`
-In Chrome dev tools, inspect static/build/bundle.[hash] .js
  - File should not be minified
  - File should have associated map file referenced at the bottom

For reference, compare this to prod (minified JS created plus map files): 
- Remove bundled JS and associated files:  `rm src/clincoded/static/build/*js*`
- Install npm packages, re-run grunt to have "production" grunt env: `bin/buildout -c buildout.cfg`
- In Chrome dev tools, inspect static/build/bundle.[hash] 
  - File should be minified
  - File should have associated map file referenced at the bottom

Note:
- Re-run  `bin/buildout -c buildout-dev.cfg` to get back to development unminified bundle.js

**To test #507:**
- Remove bundled JS and associated files:  `rm src/clincoded/static/build/*js*`
- `bin/grunt`
  - Output should include new tasks for filerev and replace:
![image](https://cloud.githubusercontent.com/assets/1878194/13756482/4eb903b4-e9dd-11e5-8caa-2397cdfb459f.png)

 - In Chrome dev tools, you should see in (index) the references to the hash-named versions of bundle.js and the css files
![image](https://cloud.githubusercontent.com/assets/1878194/13756546/9cd270bc-e9dd-11e5-9ebc-a715f93ad3fc.png)

 - Additionally you should see in static/build that the bundle.js has the hash name and static/css as well, e.g. style.395d9d2b.css, bundle.07b463da.js


**Combined test of #290 and #507:**
 - Start the site as usual. Click around. (No specific feature has changed, but check that the site behaves normally and that the icons/styles look correct)


**NOTE** The implementation could be revisited to version other resources (images, other JS, etc) but since syle.css and bundle.js are the most frequently changing between releases they seemed sufficient for now.

**NOTE2** There are other packages that deal with replacing versioned filenames, but we have a particular need to use a method like the one here because the bundle.js needs to have it's own reference to itself updated or it causes React checksum errors in console when the site is started.

**NOTE3** A small additional change to cloud-config unrelated specifically to either ticket but applicable to EC2 instances is included in PR for better diagnostics on EC2 instances. (Commands run on the server will have a `+` before them in the cloud log. [recommended by L.R.])
